offtarget/main.go

```diff
@@ -73,6 +82,141 @@ func main() {
 		}
 
 	})
+
+	http.HandleFunc("/api/2/search", func(writer http.ResponseWriter, request *http.Request) {
+		if request.Method != "POST" {
+			http.Error(writer, "This endpoint only supports POST requests", 400)
+			return
+		}
+		
+		type SearchQuery struct {
+			Queries []string
+			Radius string
+			Return_locations bool
+		}
+
+		log.Println("/api/2/search query")
+		
+		var query SearchQuery
+		
+		if request.Body == nil {
+			http.Error(writer, "Please specify search parameters (search, radius, return_locations) in request body", 400)
+			return
+		}
+
+		err := json.NewDecoder(request.Body).Decode(&query);
+		if err != nil {
+			log.Println(err.Error())
+			http.Error(writer, "Error decoding search parameters. Expected JSON in request body with: queries, list of strings; radius, string specifying search radius, return_locations, bool\n" + err.Error(), 400)
+			return
+		}
+
+		// We can only return locations if we were given an input file on disk
+		if query.Return_locations && u.Scheme != "file" {
+			http.Error(writer, "This server doesn't support return_locations", 400)
+			return
+		}
+		
+		
+		// validate search query
+		for _, protospacer := range query.Queries {
+			if len(protospacer) != 20 {
+				http.Error(writer, "queries parameter must be list of 20-mers", 400)
+				return
+			}
+		}
+
+		var radiiRe = regexp.MustCompile(`(\d+)_(\d+)_(\d+)`)
+		radii := radiiRe.FindStringSubmatch(query.Radius)
+
+		if len(radii) != 4 {
+			http.Error(writer, fmt.Sprintf("radius parameter must be 5_9_x or 5_10_x, got %s", query.Radius), 400)
+			return
+		}
+
+		lim_c5, _ := strconv.Atoi(radii[1])
+		lim_c10, _ := strconv.Atoi(radii[2])
+		lim_c20, _ := strconv.Atoi(radii[3])
+
+		if lim_c5 != 5 || (lim_c10 != 9 && lim_c10 != 10) || (lim_c20 < lim_c10) || lim_c20 > 20 {
+			http.Error(writer, fmt.Sprintf("radius parameter must be 5_9_x, with 9 <= x <= 20, or 5_10_x, with 10 <= x <= 20, but got %s", query.Radius), 400)
+			return
+		}
+
+		log.Println(fmt.Sprintf("Running query on %d protospacers with radius %d_%d_%d", len(query.Queries), lim_c5, lim_c10, lim_c20))
+
+		if !query.Return_locations {
+			for _, t := range query.Queries {
+				b := m.MatchForward(t, lim_c5, lim_c10, lim_c20)
+				// logging here is too expensive
+				//log.Printf("limits: %d %d %d targets %#v %t\n", lim_c5, lim_c10, lim_c20, t, b)
+				fmt.Fprintf(writer, "%s %t\n", t, b)
+			}
+		} else {
+			type Match struct {
```

Comment:
Some of these structs might be good to have as general structs so they aren't all buried inside of functions. I'm a fan of having nicely defined data structs at the top level.